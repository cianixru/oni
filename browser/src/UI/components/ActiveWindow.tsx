/**
 * ActiveWindow.tsx
 *
 * Helper component that is always sized and positioned around the currently
 * active window in Neovim.
 */

import * as React from "react"
import { connect } from "react-redux"

import * as Selectors from "./../Selectors"
import * as State from "./../State"

export interface IActiveWindowProps {
    pixelX: number
    pixelY: number
    pixelWidth: number
    pixelHeight: number
}

export class ActiveWindow extends React.PureComponent<IActiveWindowProps, {}> {
    public render(): JSX.Element {

        const px = (str: number): string => `${str}px`

        const style: React.CSSProperties = {
            position: "absolute",
            left: px(this.props.pixelX),
            top: px(this.props.pixelY),
            width: px(this.props.pixelWidth),
            height: px(this.props.pixelHeight),
        }

        return <div style={style}>
            {this.props.children}
        </div>
    }
}

const mapStateToProps = (state: State.IState): IActiveWindowProps => {
    const dimensions = Selectors.getActiveWindowPixelDimensions(state)
    return {
        pixelX: dimensions.x,
        pixelY: dimensions.y,
        pixelWidth: dimensions.width,
        pixelHeight: dimensions.height,
    }
}

export const ActiveWindowContainer = connect(mapStateToProps)(ActiveWindow)
