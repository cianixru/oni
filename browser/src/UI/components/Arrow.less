@import (reference) "./common.less";

.arrow {
    animation-duration: 0.3s;
    animation-delay: 0.2s;
    opacity: 0;
    animation-fill-mode: forwards;

    &.right {
        animation-name: appear-right;
    }

    &.left {
        animation-name: appear-left;
    }

    &.down {
        animation-name: appear-down;
    }

    &.up {
        animation-name: appear-up;
    }
}

@keyframes appear-down {
    from {
        transform: translateY(-4px);
        opacity: 0;
    }
    to {
        transform: translateY(0px);
        opacity: 1;
    }
}
@keyframes appear-up {
    from {
        transform: translateY(4px);
        opacity: 0;
    }
    to {
        transform: translateY(0px);
        opacity: 1;
    }
}
@keyframes appear-left {
    from {
        transform: translateX(4px);
        opacity: 0;
    }
    to {
        transform: translateX(0px);
        opacity: 1;
    }
}
@keyframes appear-right {
    from {
        transform: translateX(-4px);
        opacity: 0;
    }
    to {
        transform: translateX(0px);
        opacity: 1;
    }
}
