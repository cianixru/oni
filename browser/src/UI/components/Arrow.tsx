/**
 * Arrow.tsx
 *
 * Simple 'up' or 'down' arrow component
 */
import * as React from "react"

import styled from "styled-components"
import { withProps } from "./common"

export enum ArrowDirection {
    Up = 0,
    Down = 1,
    Left = 2,
    Right = 3,
}

export interface IArrowProps {
    size: number
    color: string
    direction: ArrowDirection
}

const transparentBorder = (props: IArrowProps) => `${props.size * 0.8}px solid transparent`
const solidBorder = (props: IArrowProps) => `${props.size}px solid ${props.color}`

function arrowCSS(props: IArrowProps): string {
    switch (props.direction) {
        case ArrowDirection.Up:
            return `
        width: 0px;
        height: 0px;
        border-left: ${transparentBorder(props)};
        border-right: ${transparentBorder(props)};
        border-bottom: ${solidBorder(props)};
        `
        case ArrowDirection.Down:
            return `
        width: 0px;
        height: 0px;
        border-left: ${transparentBorder(props)};
        border-right: ${transparentBorder(props)};
        border-top: ${solidBorder(props)};
        `
        case ArrowDirection.Left:
            return `
        width: 0px;
        height: 0px;
        border-top: ${transparentBorder(props)};
        border-right: ${solidBorder(props)};
        border-bottom: ${transparentBorder(props)};
        `
        case ArrowDirection.Right:
            return `
        width: 0px;
        height: 0px;
        border-top: ${transparentBorder(props)};
        border-left: ${solidBorder(props)};
        border-bottom: ${transparentBorder(props)};
        `
        default:
            return ``
    }
}

export const Arrow = withProps<IArrowProps>(styled.div)`
    ${props => arrowCSS(props)}
    `

export const preview = () => {
    return <Arrow size={100} color="yellow" direction={ArrowDirection.Left} />
}
