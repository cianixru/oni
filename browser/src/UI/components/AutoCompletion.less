
.autocompletion {
    background-color: rgb(40, 40, 40);
    border: 1px solid gray;
    box-shadow: 4px 4px rgba(0, 0, 0, 0.2);
    color: rgb(200, 200, 200);
    max-width: 800px;
    max-height: 300px;
    overflow-x: hidden;
    overflow-y: auto;
    animation-name: appear;
    animation-duration: 0.1s;

    .entry {
        &.selected {
            background-color: rgb(60, 60, 60);

            .main {
                opacity: 1;

                .icon {
                    border-left: 1px solid white;
                }
            }
        }

        .main {
            transition: opacity 1s;
            opacity: 0.8;
            display: flex;
            flex-direction: row;
            align-items: center;

            .icon {
                padding: 4px;
                width: 20px;
                flex: 0 0 auto;
                color: white;

                i {
                    margin-left: 4px;
                    margin-right: 4px;
                    font-size: 10px;
                }
            }
        }

        .label {
            flex: 1 0 auto;
            min-width:100px;
            margin-left: 8px;

            .highlight {
                font-weight: bold;
                font-style: italic;
                color: rgb(50, 200, 255);
            }
        }

        .detail {
            flex: 0 1 auto;
            min-width: 50px;
            text-align: right;

            text-overflow: ellipsis;
            overflow: hidden;
            white-space: nowrap;
            margin-left: 16px;

            color: rgba(100, 100, 100, 0.6);

            font-size: 11px;
        }

        .documentation {
            font-size: 10px;
            color: rgb(50, 200, 255);
        }
    }
}

@keyframes appear {
    from {opacity: 0;}
    to {opacity: 1;}
}

