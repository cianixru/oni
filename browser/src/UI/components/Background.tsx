import * as React from "react"

import { connect } from "react-redux"
import * as State from "./../State"

export interface IBackgroundProps {
    backgroundColor: string
    backgroundImageUrl: string
    backgroundImageSize: string
    backgroundOpacity: number
}

export class BackgroundView extends React.PureComponent<IBackgroundProps, void> {
    public render(): JSX.Element {
        const coverStyle = {
            backgroundColor: this.props.backgroundColor,
            opacity: this.props.backgroundOpacity,
        }

        return <div>
            <BackgroundImageView {...this.props} />
            <div className="background-cover" style={coverStyle}></div>
        </div>
    }
}

export const BackgroundImageView = (props: IBackgroundProps) => {

    if (props.backgroundImageUrl) {
        const imageStyle = {
            backgroundImage: "url(" + props.backgroundImageUrl + ")",
            backgroundSize: props.backgroundImageSize || "cover",
        }

        return <div className="background-image" style={imageStyle}></div>
    } else {
        return null
    }
}

const mapStateToProps = (state: State.IState): IBackgroundProps => {
    const conf = state.configuration
    return {
        backgroundColor: state.backgroundColor,
        backgroundImageUrl: State.readConf(conf, "editor.backgroundImageUrl"),
        backgroundImageSize: State.readConf(conf, "editor.backgroundImageSize"),
        backgroundOpacity: State.readConf(conf, "editor.backgroundOpacity"),
    }
}

export const Background = connect(mapStateToProps)(BackgroundView)
