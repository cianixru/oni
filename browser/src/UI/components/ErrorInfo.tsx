import * as React from "react"
import * as types from "vscode-languageserver-types"

import styled from "./common"
import { ErrorIcon } from "./Error"

import { getColorFromSeverity } from "./../../Services/Diagnostics"

export interface IErrorInfoProps {
    hasQuickInfo: boolean
    errors: types.Diagnostic[]
}

export const DiagnosticMessage = styled.span`
    margin-left: 1em;
`

type StyleProps = Pick<IErrorInfoProps, "hasQuickInfo">

const DiagnosticContainer = styled<StyleProps, "div">("div")`
    user-select: none;
    cursor: default;
    border-bottom: ${p => (p.hasQuickInfo ? `1px solid ${p.theme["toolTip.border"]}` : "")};
`

export const Diagnostic = styled.div`
    margin: 8px;
    display: flex;
    flex-direction: row;
`

/**
 * Helper component to render errors in the QuickInfo bubble
 */
export const ErrorInfo = (props: IErrorInfoProps) => {
    return (
        props.errors && (
            <DiagnosticContainer hasQuickInfo={props.hasQuickInfo}>
                {props.errors.map((e, idx) => (
                    <Diagnostic key={e.code + e.message + e.source + idx}>
                        <ErrorIcon color={getColorFromSeverity(e.severity)} />
                        <DiagnosticMessage data-id="diagnostic-message">
                            {e.message}
                        </DiagnosticMessage>
                    </Diagnostic>
                ))}
            </DiagnosticContainer>
        )
    )
}
