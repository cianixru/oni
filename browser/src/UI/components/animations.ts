import { keyframes } from "./common"

export const fadeInAndDown = keyframes`
    from {
        opacity: 0;
    }
    to {
        opacity: 1;
    }
`
