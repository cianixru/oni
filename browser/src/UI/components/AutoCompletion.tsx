import * as React from "react"
import { connect } from "react-redux"

import * as _ from "lodash"

import { Icon } from "./../Icon"
import { IState /*, AutoCompletionInfo */ } from "./../State"
import { HighlightText } from "./HighlightText"

export interface IAutoCompletionProps {
    visible: boolean
    x: number
    y: number
    base: string
    entries: Oni.Plugin.CompletionInfo[]
    selectedIndex: number
}

require("./AutoCompletion.less") // tslint:disable-line no-var-requires

export class AutoCompletion extends React.Component<IAutoCompletionProps, void> {

    public render(): null | JSX.Element {

        if (!this.props.visible) {
            return null
        }

        const containerStyle = {
            position: "absolute",
            top: this.props.y.toString() + "px",
            left: this.props.x.toString() + "px",
        }

        if (this.props.entries.length === 0) {
            return null
        }

        if (this.props.entries.length === 1
            && this.props.entries[0].label === this.props.base) {
            return null
        }

        const firstTenEntries = _.take(this.props.entries, 10)

        const entries = firstTenEntries.map((s, i) => {

            const isSelected = i === this.props.selectedIndex

            return <AutoCompletionItem {...s} isSelected={isSelected} base={this.props.base} />
        })

        return (<div style={containerStyle} className="autocompletion">
            {entries}
        </div>)
    }
}

export interface IAutoCompletionItemProps extends Oni.Plugin.CompletionInfo {
    base: string
    isSelected: boolean
}

export class AutoCompletionItem extends React.Component<IAutoCompletionItemProps, void> {
    public render(): JSX.Element {

        let className = "entry"
        if (this.props.isSelected) {
            className += " selected"
        }

        const detailToShow = this.props.isSelected ? this.props.detail : ""
        const documentation = this.props.isSelected ? this.props.documentation : ""

        const highlightColor = this.props.highlightColor || this._getDefaultHighlightColor(this.props.kind as any) // FIXME: undefined

        const iconContainerStyle = {
            backgroundColor: highlightColor,
        }

        return <div className={className}>
            <div className="main">
                <span className="icon" style={iconContainerStyle}>
                    <AutoCompletionIcon kind={this.props.kind as any /* FIXME: undefined */} />
                </span>
                <HighlightText className="label" highlightClassName="highlight" highlightText={this.props.base} text={this.props.label} />
                <span className="detail">{detailToShow}</span>
            </div>
            <div className="documentation">{documentation}</div>
        </div>
    }

    private _getDefaultHighlightColor(_kind: string): string {
        // TODO: Extend this logic for better defaults per kind
        return "rgb(32, 232, 38)"
    }
}

export interface IAutoCompletionIconProps {
    kind: string
}

export class AutoCompletionIcon extends React.Component<IAutoCompletionIconProps, void> {

    public render(): null | JSX.Element {

        switch (this.props.kind) {
            case "let":
                return <Icon name="wrench" />
            case "interface":
                return <Icon name="plug" />
            case "alias":
                return <Icon name="id-badge" />
            case "const":
                return <Icon name="lock" />
            case "class":
                return <Icon name="cube" />
            case "type":
                return <Icon name="sitemap" />
            case "directory":
                return <Icon name="folder" />
            case "file":
            case "script":
                return <Icon name="file" />
            case "var":
            case "property":
            case "parameter":
                // Closed cube?
                return <Icon name="code" />
            case "module":
            case "external module name":
                return <Icon name="cubes" />
            case "method":
            case "function":
                return <Icon name="cog" />
            case "keyword":
                return <Icon name="key" />
            case "text":
                return <Icon name="align-justify" />
            case "warning":
            case "$warning":
                return <Icon name="exclamation-triangle" />
            default:
                return !this.props.kind ? null : <span>`?${this.props.kind}?`</span>
        }
    }
}

const mapStateToProps = (state: IState) => {
    if (!state.autoCompletion) {
        return {
            visible: false,
            x: state.cursorPixelX,
            y: state.cursorPixelY + state.fontPixelHeight,
            base: "",
            entries: [],
            selectedIndex: 0,
        }
    } else {
        const ret: IAutoCompletionProps = {
            visible: true,
            x: state.cursorPixelX,
            y: state.cursorPixelY + state.fontPixelHeight,
            base: state.autoCompletion.base,
            entries: state.autoCompletion.entries,
            selectedIndex: state.autoCompletion.selectedIndex,
        }
        return ret
    }
}

export const AutoCompletionContainer = connect(mapStateToProps)(AutoCompletion)
