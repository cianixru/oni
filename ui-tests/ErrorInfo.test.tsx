import * as React from "react"
import { shallow } from "enzyme"
import { Diagnostic } from "vscode-languageserver-types"

import { ErrorInfo, DiagnosticMessage } from "../browser/src/UI/components/ErrorInfo"

describe("<ErrorInfo />", () => {
    const errors: Diagnostic[] = [
        {
            range: {
                start: {
                    line: 0,
                    character: 0,
                },
                end: {
                    line: 0,
                    character: 10,
                },
            },
            severity: 2,
            code: 5,
            source: "test",
            message: "an error here",
        },
    ]

    it("Should Render without crashing", () => {
        const wrapper = shallow(<ErrorInfo hasQuickInfo={true} errors={errors} />)
        expect(wrapper.length).toBe(1)
    })
    it("Should Not Render without Errors", () => {
        const wrapper = shallow(<ErrorInfo hasQuickInfo={false} errors={null} />)
        expect(wrapper.children().length).toBe(0)
    })
    it("Should render the correct diagnostic message", () => {
        const wrapper = shallow(<ErrorInfo hasQuickInfo={true} errors={errors} />)
        expect(wrapper.dive().find("[data-id='diagnostic-message']").length).toEqual(1)
        expect(
            wrapper
                .dive()
                .find("[data-id='diagnostic-message']")
                .dive()
                .text(),
        ).toBe(errors[0].message)
    })

    it("Should render the correct number of diagnostics", () => {
        const wrapper = shallow(
            <ErrorInfo hasQuickInfo={true} errors={[...errors, ...errors, ...errors]} />,
        )
        expect(wrapper.dive().find("[data-id='diagnostic-message']").length).toBe(3)
    })

    it("Should match the last snapshot on record unless a purposeful change was made", () => {
        const wrapper = shallow(
            <ErrorInfo hasQuickInfo={true} errors={[...errors, ...errors, ...errors]} />,
        )
        expect(wrapper).toMatchSnapshot()
    })
})
