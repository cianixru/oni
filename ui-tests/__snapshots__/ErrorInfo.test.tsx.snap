// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorInfo /> Should match the last snapshot on record unless a purposeful change was made 1`] = `
<styled.div
  hasQuickInfo={true}
>
  <styled.div
    key="5an error heretest0"
  >
    <Component
      color="yellow"
    />
    <styled.span
      data-id="diagnostic-message"
    >
      an error here
    </styled.span>
  </styled.div>
  <styled.div
    key="5an error heretest1"
  >
    <Component
      color="yellow"
    />
    <styled.span
      data-id="diagnostic-message"
    >
      an error here
    </styled.span>
  </styled.div>
  <styled.div
    key="5an error heretest2"
  >
    <Component
      color="yellow"
    />
    <styled.span
      data-id="diagnostic-message"
    >
      an error here
    </styled.span>
  </styled.div>
</styled.div>
`;
